kctf-operator/pkg/controller/challenge/challenge_controller.go

```diff
@@ -0,0 +1,183 @@
+		}
+
+		// Error reading the object - requeue the request.
+		reqLogger.Error(err, "Failed to get Challenge")
+		return reconcile.Result{}, err
+	}
+
+	// Check if the deployment already exists, if not create a new one
+	found := &appsv1.Deployment{}
+	err = r.client.Get(context.TODO(), types.NamespacedName{Name: challenge.Name, Namespace: challenge.Namespace}, found)
+
+	// Set default values not configured by kubebuilder
+	SetDefaultValues(challenge)
+
+	// Just enters here if it's a new deployment
+	if err != nil && errors.IsNotFound(err) {
+		// Define a new deployment
+		dep := r.deploymentForChallenge(challenge)
+		reqLogger.Info("Creating a new Deployment", "Deployment.Namespace", dep.Namespace, "Deployment.Name", dep.Name)
+		err = r.client.Create(context.TODO(), dep)
+
+		if err != nil {
+			reqLogger.Error(err, "Failed to create new Deployment", "Deployment.Namespace", dep.Namespace, "Deployment.Name", dep.Name)
+			return reconcile.Result{}, err
+		}
+
+		// Deployment created successfully - return and requeue
+		return reconcile.Result{Requeue: true}, nil
+
+	} else if err != nil {
+		reqLogger.Error(err, "Failed to get Deployment")
+		return reconcile.Result{}, err
+	}
+
+	// Creates the service if it doesn't exist
+	// Check existence of the service:
+	serviceFound := &corev1.Service{}
+	err = r.client.Get(context.TODO(), types.NamespacedName{Name: challenge.Name, Namespace: challenge.Namespace}, serviceFound)
+
+	// Just enter here if the service doesn't exist yet:
+	if err != nil && errors.IsNotFound(err) && challenge.Spec.Network.Public == true {
+		// Define a new service if the challenge is public
+		serv := r.serviceForChallenge(challenge)
+		reqLogger.Info("Creating a new Service", "Service.Namespace", serv.Namespace, "Service.Name", serv.Name)
+		err = r.client.Create(context.TODO(), serv)
+
+		if err != nil {
+			reqLogger.Error(err, "Failed to create new Service", "Deployment.Service", serv.Namespace, "Service.Name", serv.Name)
+			return reconcile.Result{}, err
+		}
+
+		// Service created successfully - return and requeue
+		return reconcile.Result{Requeue: true}, nil
+
+		// When service exists and public is changed to false
+	} else if err == nil && challenge.Spec.Network.Public == false {
+		reqLogger.Info("Deleting the Service", "Service.Namespace", serviceFound.Namespace, "Service.Name", serviceFound.Name)
+		err = r.client.Delete(context.TODO(), serviceFound)
+
+		if err != nil {
+			reqLogger.Error(err, "Failed to erase Service", "Deployment.Service", serviceFound.Namespace, "Service.Name", serviceFound.Name)
+			return reconcile.Result{}, err
+		}
+
+		// Service deleted successfully - return and requeue
+		return reconcile.Result{}, err
+	}
+
+	// Ensure that the configurations in the CR are followed - Checks done everytime the CR is updated
+	// change says if something in the configurations was different from what was found in the deploymeny
+	change := CheckConfigurations(challenge, found)
```

Comment:
When I took the example of an operator, what we had in the challenge-controller was the deployment and, after that, a code that checked if the quantity of replicas was right. What I have done was to put the deployment and this code that checks if the configurations are well set in other files. Also, I created the service and created a function that set defaults values in variables that can't be set to default in CRD.

The deployment is created only once and the service as well if the variable public doesn't change. The only call that it's in all reconcile function calls is the CheckConfigurations, which changes the deployment/service if the author changes something in the CR. The other function that set default values is also called in all reconcile calls, because, when the author changes the CR, we have to pass them again.

Reply:
@sroettger would changing from check to a better name that describes what it does fix this concern?

Reply:
as above, changing the name would resolve this